python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,303 @@
+        for key, value in tags.items():
+            key = key.lower().strip()
+            if key.startswith("rikai.option."):
+                sub_len = len("rikai.option.")
+                options[key[sub_len:]] = value
+        if len(options) > 0:
+            spec["options"] = options
+
+        # uri (to the model artifact)
+        if "rikai.model.artifact_path" in tags:
+            spec["model"]["uri"] = os.path.join(
+                "runs:/",
+                self._run.info.run_id,
+                tags["rikai.model.artifact_path"],
+            )
+        return spec
+
+
+def codegen_from_run(
+    spark: SparkSession,
+    run: mlflow.entities.Run,
+    tracking_uri: str,
+    name: Optional[str] = None,
+    options: Optional[Dict[str, str]] = None,
+) -> str:
+    """Generate code from an MLFlow runid
+
+    Parameters
+    ----------
+    spark : SparkSession
+        A live spark session
+    run : mlflow.entities.Run
+        the mlflow run that produced the model we want to register
+    tracking_uri: str
+        The mlflow tracking uri
+    name : str
+        The name of the model in the catalog
+    options : dict
+        Optional parameters passed to the model.
+
+    Returns
+    -------
+    str
+        Spark UDF function name for the generated data.
+    """
+    try:
+        spec = MlflowModelSpec(run, tracking_uri, options=options)
+    except Exception as e:
+        to_spec_msg = (
+            "Could not create well-formed ModelSpec from run {}."
+            "Check MLFlowModelSpec._run_to_spec_dict"
+        ).format(run.run_id)
+        raise SpecError(to_spec_msg) from e
+    udf = udf_from_spec(spec)
+    return register_udf(spark, udf, name)
+
+
+def log_model(
+    model: Any,
+    artifact_path: str,
+    flavor: str,
+    schema: str,
+    pre_processing: Optional[str] = None,
+    post_processing: Optional[str] = None,
+    registered_model_name: Optional[str] = None,
+    **kwargs,
+):
+    """Convenience function to log the model with information needed by rikai.
+    This should be called during training when the model artifact is produced.
+
+
+    Parameters
+    ----------
+    model: Any
+        The model artifact object
+    artifact_path: str
+        The relative (to the run) artifact path
+    flavor: str
+        A valid Mlflow flavor (e.g., "pytorch")
+    schema: str
+        Output schema (pyspark DataType)
+    pre_processing: str, default None
+        Full python module path of the pre-processing transforms
+    post_processing: str, default None
+        Full python module path of the post-processing transforms
+    registered_model_name: str, default None
+        Model name in the mlflow model registry
+    kwargs: dict
+        Passed to `mlflow.<flavor>.log_model`
+
+    Examples
+    --------
+    # import mlflow.pytorch  <= vanilla mlflow usage
+    from rikai.spark.sql.codegen.mlflow_registry import log_model
```

Comment:
we might want to move this to `rikai.mlflow` package if we expect users to use this API?  this `rikai.spark.sql.codegen` is long and sounds like internal package.

Reply:
yeah, that's not a bad idea